private/buf/buflsp/file.go

```diff
@@ -736,6 +743,41 @@ func findImportable(
 	return imports, nil
 }
 
+// getWorkspaceWithLocalPaths checks each proto file in a workspace and ensures they all have
+// a LocalPath. If not, then it refreshes the workspace. As outlined on [storage.ObjectInfo],
+// LocalPath will not always be present if the file did not originate from disk. If an import's
+// module data was fetched remotely, then it will not have a path on disk to the cache. By
+// refreshing the module, and rebuilding the workspace, we populate the module data from the cache.
+func getWorkspaceWithLocalPaths(ctx context.Context, fileName string, lsp *lsp) (bufworkspace.Workspace, error) {
+	workspace, err := lsp.controller.GetWorkspace(ctx, fileName)
+	if err != nil {
+		return nil, err
+	}
+	var refreshWorkspace bool
+	for _, module := range workspace.Modules() {
+		if err := module.WalkFileInfos(ctx, func(fileInfo bufmodule.FileInfo) error {
+			if fileInfo.FileType() == bufmodule.FileTypeProto && fileInfo.LocalPath() == "" {
+				return fmt.Errorf("file %q no local path: %w", fileInfo.Path(), errFileInfoNoLocalPath)
+			}
+			return nil
+		}); err != nil {
+			if errors.Is(err, errFileInfoNoLocalPath) {
+				lsp.logger.Info(fmt.Sprintf("re-creating workspace: %v", err))
+				refreshWorkspace = true
+				break
+			}
+			return nil, err
+		}
+	}
+	if refreshWorkspace {
+		workspace, err = getWorkspaceWithLocalPaths(ctx, fileName, lsp)
```

Comment:
I don't understand what is going on here (which means either it needs code clarification, or this function needs to be refactored). It appears you're making a recursive call with no clear base case that will result in recursion terminating. Moreover, each recursive call is O(files).

Reply:
I can clarify the comment at the top-level of the function and in-line more information, however this is an accurate description of what is going on. We are trying to ensure that the workspace that is built, for all imports, we can resolve a `LocalPath`. And the only way to check for that is to iterate through each file, there are no other guarantees otherwise.

The exit condition is where all files have a `LocalPath`, which should be the state we converge to once all imports are being resolved through the cache.

This is an expensive operation as it is implemented because of the conditions set by the LSP. In practice, we should have a fully populated cache and module data on-disk fairly quickly, since we cache all module data.

The other option we have is to change the semantics of `ModuleDataProvider`, such that if we resolve remote module data and put those values in the cache after, then we return module data with the cached information.